README.md
<div align="center">
  <img src="./logo.svg" />
</div>

effepi [![Build Status](https://travis-ci.org/OctoD/effepi.svg?branch=master)](https://travis-ci.org/OctoD/effepi)
======

Fun functional programming with pipelinable functions.

## What effepi is

Effepi is a functional way to enqueue and use different functions. You can put your functions in a pipeline and resolve them both asynchronously or synchronously, creating new async functions or sync functions.

## Install

#### Npm 

```bash
npm i effepi
```

#### Yarn 

```bash
yarn add effepi
```

## Index

- [effepi ![Build Status](https://travis-ci.org/OctoD/effepi)](#effepi-build-statushttpstravis-ciorgoctodeffepi)
  - [What effepi is](#what-effepi-is)
  - [Install](#install)
      - [Npm](#npm)
      - [Yarn](#yarn)
  - [Index](#index)
  - [Install](#install-1)
  - [How to use it](#how-to-use-it)
      - [pipeline context](#pipeline-context)
  - [Functions](#functions)
      - [Array functions](#array-functions)
          - [applyEach](#applyeach)
          - [applyEachSync](#applyeachsync)
          - [concat](#concat)
          - [filter](#filter)
          - [filterWith](#filterwith)
          - [find](#find)
          - [findExact](#findexact)
          - [join](#join)
          - [length](#length)
          - [nth](#nth)
          - [reverse](#reverse)
      - [Boolean functions](#boolean-functions)
          - [F](#f)
          - [T](#t)
          - [inverse](#inverse)
      - [Math functions](#math-functions)
          - [add](#add)
          - [changeSign](#changesign)
          - [decrement](#decrement)
          - [divideBy](#divideby)
          - [increment](#increment)
          - [multiplyBy](#multiplyby)
          - [negative](#negative)
          - [positive](#positive)
          - [pow](#pow)
          - [root](#root)
          - [subtract](#subtract)
          - [takeBetween](#takebetween)
          - [takeGreater](#takegreater)
          - [takeGreaterThan](#takegreaterthan)
          - [takeLower](#takelower)
          - [takeLowerThan](#takelowerthan)
          - [takeOuter](#takeouter)
      - [Logical operators functions](#logical-operators-functions)
          - [createSwitch](#createswitch)
          - [fold](#fold)
          - [ifElse](#ifelse)
      - [Object functions](#object-functions)
          - [exclude](#exclude)
          - [hasProperty](#hasproperty)
          - [keys](#keys)
          - [maybe](#maybe)
          - [merge](#merge)
          - [pick](#pick)
      - [Misc functions](#misc-functions)
          - [apply](#apply)
          - [applySync](#applysync)
          - [put](#put)
          - [safeCall](#safecall)
          - [useCallValue](#usecallvalue)
          - [useValue](#usevalue)
      - [String functions](#string-functions)
          - [camelCase](#camelcase)
          - [chars](#chars)
          - [concat](#concat-1)
          - [includes](#includes)
          - [length](#length-1)
          - [lowercase](#lowercase)
          - [pascalCase](#pascalcase)
          - [repeat](#repeat)
          - [replaceAll](#replaceall)
          - [toBinaryArray](#tobinaryarray)
          - [uppercase](#uppercase)
      - [Type functions](#type-functions)
          - [exactTypeOf](#exacttypeof)
          - [ofType](#oftype)
          - [toArray](#toarray)
          - [toBoolean](#toboolean)
          - [toDate](#todate)
          - [toNumber](#tonumber)
          - [toSet](#toset)
          - [toString](#tostring)
  - [Examples](#examples)
  - [Contributing](#contributing)
  - [Licence](#licence)

## Install

```bash
npm i pipe

# or 

yarn add pipe
```

## How to use it

```ts
import { pipe, math, misc } from 'effepi';

export const calculateVat = (vatPercentage: number) =>
  pipe<number, number>(misc.useCallValue())
    .pipe(math.divideBy(100))
    .pipe(math.multiplyBy(vatPercentage))
    .toSyncFunction();

vatCalculator(22)(100); // 22
vatCalculator(22)(1285); // 282.7
```

To create a pipeline, use the `pipe` function. This function can be used with any function, althought is recommended if you are building a function from a pipeline to use the `useCallValue` function.

```ts
const p = pipe(useCallValue()).toSyncFunction();

p(10) // returns 10
p('hello world') // returns 'hello world'
```

A pipeline can be memoized, it is useful when you have to call a pipeline a lot of times.

```ts
const test = pipe(useCallValue(), true)
  .pipe(add(10))
  .pipe(multiplyBy(2))
  .toSyncFunction();

test(1) // returns 22
test(1) // returns the cached previous result
test(2) // returns 24
test(2) // returns the cached previous result
test(2) // returns the cached previous result
```

#### pipeline context

Each function passed in a pipeline can use the previous value and can access to the current pipeline context. A passed context can be enriched with a `mutation` function, which is the only way to mutate the pipeline from the inside of a function.

```ts
const aFunctionUsingContext = (previousValue, context) => {
  // mutations are always executed after the current function
  context.mutate = () => {
    const pipeline = [
      (arg: number) => {
        const isEven = arg % 2 === 0;

        console.log(isEven);
        
        return isEven;
      },
    ];

    return { pipeline };
  };
  console.log(previousValue);
  return previousValue;
};

const isEven = pipe(useCallValue())
  .pipe(aFunctionUsingContext) // uses console.log for context callValue
  .toSyncFunction(); // returns true if the number is even

isEven(10) // logs 10, logs true, returns true
```

A context has also the `call` method, which can be used to invoke a function with
the previous value as argument.

```ts
pipe(useCallValue())
  .pipe(
    (value, context) => context.call(add(10))
  )
).resolveSync(0); // returns 10
```

## Functions

#### Array functions

Array functions are under the `array` module.

```ts
import { array } from 'effepi';
```

###### applyEach

Applies the result of a pipeline to each element in an array of another pipeline.

Note: invoking a pipeline using this function with a sync method will throw an error.

```ts
const doMath = pipe(useCallValue())
  .pipe(sum(2))
  .pipe(multiplyBy(3));

pipe(useCallValue())
  .pipe(applyEach(doMath))
  .resolve([10, 20, 30]) // Promise([36, 66, 96])
```

###### applyEachSync

Is the same of applyEach, except it does not work with async functions

Note: invoking a pipeline using this function with an async method will throw an error.

```ts
const doMath = pipe(useCallValue())
  .pipe(sum(2))
  .pipe(multiplyBy(3));

pipe(useCallValue())
  .pipe(applyEach(doMath))
  .resolveSync([10, 20, 30]) // [36, 66, 96]
```

###### concat

Concatenates previous value with another array.

If the previous value is not an array an error will be thrown

```ts
pipe(useCallValue())
  .pipe(concat([4,5,6]))
  .resolveSync([1,2,3]) // [1,2,3,4,5,6]
```

###### filter

Filters the previous value with a given callback. The callback must return a boolean value. 

If the previous value is not an array an error will be thrown

```ts
pipe(useCallValue())
  .pipe(filter(a => a > 2))
  .resolveSync([1,2,3]) // [3]
```

###### filterWith

Filters the previous value with a given value.

If the previous value is not an array an error will be thrown

```ts
pipe(useCallValue())
  .pipe(filterWith(2))
  .resolveSync([1,2,3,2,3,2]) // [2,2,2]
```

###### find

Finds a value as specified by a find callback. The callback must return `true`.

If the value is not found, the pipe will return an `undefined`

If the previous value is not an array an error will be thrown

```ts
pipe(useCallValue())
  .pipe(find((arg: number) => arg === 1))
  .resolveSync([1,2,3,2,3,2]) // 1
```

###### findExact

Finds an exacts value.

If the value is not found, the pipe will return an `undefined`

If the previous value is not an array an error will be thrown

```ts
pipe(useCallValue())
  .pipe(find(1))
  .resolveSync([1,2,3,2,3,2]) // 1
```

###### join

Joins the previous value with a given char. 

If the previous value is not an array an error will be thrown

```ts
pipe(useCallValue())
  .pipe(join('*'))
  .resolveSync([1,2,3]) // '1*2*3'
```

###### length

Returns the length of the previous value.

If the previous value is not an array an error will be thrown

```ts
pipe(useCallValue())
  .pipe(length())
  .resolveSync([1,2,3]) // 3
```

###### nth

Returns the nth element in the previous value.

If the previous value is not an array an error will be thrown

```ts
pipe(useCallValue())
  .pipe(nth(3))
  .resolveSync([0, 2, 5, 12, 24]) // 12
```

###### reverse

Reverses the previous value. 

If the previous value is not an array an error will be thrown

```ts
pipe(useCallValue())
  .pipe(reverse())
  .resolveSync([1,2,3]) // [3,2,1]
```

#### Boolean functions

Boolean functions are under the `boolean` module.

```ts
import { boolean } from 'effepi';
```

###### F

Puts a `false` value.

```ts
pipe(F()).resolveSync(undefined) // false
```

###### T

Puts a `true` value.

```ts
pipe(T()).resolveSync(undefined) // true
```

###### inverse

Inverts previous value. 

Previous value must be boolean.

```ts
pipe(useCallValue()).pipe(inverse()).resolveSync(true) // false
pipe(useCallValue()).pipe(inverse()).resolveSync(false) // true
```

#### Math functions

Math functions are under the `math` module.

```ts
import { math } from 'effepi';
```

###### add

Sums a value to the previous one. 

```ts
pipe(useCallValue()).pipe(add(1)).resolveSync(10) // 11
```

###### changeSign

Changes previous value sign, from positive to negative and vice-versa.

```ts
pipe(put(-123)).pipe(changeSign()) // 123
```

###### decrement

Decrements the previous value by one.

```ts
pipe(useCallValue()).pipe(decrement()).resolve(44) // 41
```

###### divideBy

Divides the previous value by the passed one.

```ts
pipe(useCallValue()).pipe(divideBy(2)).resolve(44) // 22
```

###### increment

Increments the previous value by one.

```ts
pipe(useCallValue()).pipe(increment()).resolve(44) // 45
```

###### multiplyBy

Multiplies the previous value by the given one

```ts
pipe(useCallValue()).pipe(multiplyBy(2)).resolve(44) // 88
```

###### negative

Converts previous' value from positive sign to negative unless is already negative.

```ts
pipe(useCallValue()).pipe(negative()).resolve(44) // -44
pipe(useCallValue()).pipe(negative()).resolve(-12) // -12
```

###### positive

Converts previous' value from negative sign to positive unless is already positive.

```ts
pipe(useCallValue()).pipe(positive()).resolve(44) // 44
pipe(useCallValue()).pipe(positive()).resolve(-12) // 12
```

###### pow

Raises to power the previous value by a given exponent

```ts
pipe(useCallValue()).pipe(pow(4)).resolve(2) // 16
pipe(useCallValue()).pipe(pow(2)).resolve(-2) // 4
```

###### root

Extracts the root of the previous value by a given exponent

```ts
pipe(useCallValue()).pipe(root(2)).resolve(4) // 2
pipe(useCallValue()).pipe(root(2)).resolve(9) // 3
```

###### subtract

Subtracts the previous value by the given one

```ts
pipe(useCallValue()).pipe(subtract(2)).resolve(9) // 7
```

###### takeBetween

Takes all numbers in a number array between two values (inclusive)

```ts
pipe(useCallValue()).pipe(takeBetween(5, 7)).resolve([4, 5, 6, 7, 8]) // [5, 6, 7]
```

###### takeGreater

Returns the greater number in a number array

```ts
pipe(useCallValue()).pipe(takeGreater()).resolve([4, 5, 44, 6, 7, 8]) // 44
```

###### takeGreaterThan

Takes all numbers in a number array greater than a given value. 

```ts
pipe(useCallValue()).pipe(takeGreaterThan(8)).resolve([4, 5, 44, 6, 7, 8]) // [44]
```

This function accepts a second parameter (boolean) to include also the same value

```ts
pipe(useCallValue()).pipe(takeGreaterThan(8, true)).resolve([4, 5, 44, 6, 7, 8]) // [8, 44]
```

###### takeLower

Returns the lower number in a number array

```ts
pipe(takeLower()).pipe(takeGreater()).resolve([4, 5, 44, 6, 7, 8]) // 4
```

###### takeLowerThan

Takes all numbers in a number array lower than a given value. 

```ts
pipe(useCallValue()).pipe(takeLowerThan(5)).resolve([4, 5, 44, 6, 7, 8]) // [4]
```

This function accepts a second parameter (boolean) to include also the same value

```ts
pipe(useCallValue()).pipe(takeLowerThan(8, true)).resolve([4, 5, 44, 6, 7, 8]) // [4, 5, 6, 7, 8]
```

###### takeOuter

Takes all numbers in a number array lower than the first passed value or greater than the second passed value. Matching elements are discarded.

```ts
pipe(useCallValue()).pipe(takeOuter(5, 10)).resolveSync([3,4,5,10,11]) // [3, 4, 11]
```


#### Logical operators functions

Logical operators functions are under the `logical` module.

```ts
import { logical } from 'effepi';
```

###### createSwitch

Is the same of the `switch` construct.

```ts
// silly example: checking if a string is inside an array of strings
const cities = pipe(useCallValue())
  .pipe(
    createSwitch(
      createSwitchDefault('City not found'),
      createSwitchOption('Munich', 'Beerfest!'),
      createSwitchOption('Rome', 'We love carbonara'),
      createSwitchOption('London', 'God save the queen'),
    )
  )
  .toSyncFunction();

cities('Munich') // 'Beerfest!'
cities('Rome') // 'We love carbonara'
cities('London') // 'God save the queen'
cities('Casalpusterlengo') // 'City not found'
```

To create the default, you can use the `createSwitchDefault` method, whilst using `createSwitchOption` you can add a switch case.

###### fold

This function is the same of the `if/else` statement. 

It takes two arguments, the `left` and the `right` part. The left part is intended as the `false` comparison result, while the `right` is the `true`.

```ts
const smsLengthCheck = pipe(useCallValue())
  .pipe(isGreaterThan(144))
  .pipe(fold('', 'Maximum character'))
  .toSyncFunction();

smsLengthCheck('lorem') // ''
smsLengthCheck('lorem'.repeat(2000)) // 'Maximum character'
```

###### ifElse

This function works like the `if/else` statement. 

It requires three arguments:

* a Condition (a `function` which returns a `boolean`)
* a Left, which can be a value or another `pipe`. If is a pipe, it will be resolved using the context's async/sync flow
* a Right, which can be a value or another `pipe`. If is a pipe, it will be resolved using the context's async/sync flow

```ts
const simple = pipe(useCallValue())
  .pipe(
    logical.ifElse(
      (arg: number) => arg > 5,
      'lower than 5',
      'greater than 5'
    )
  )
  .toSyncFunction();

const complex = pipe(useCallValue())
  .pipe(
    logical.ifElse(
      (arg: number) => arg > 5, 
      pipe(useCallValue()).pipe(math.pow(2)), 
      pipe(useCallValue()).pipe(math.divideBy(2)),
    )
  ).toSyncFunction();

simple(4) // lower than 5
simple(10) // greater than 5
complex(4) // 14
complex(10) // 5
```

#### Object functions

Object functions are under the `object` module.

```ts
import { object } from 'effepi';
```

###### exclude

Returns previous value except for the given keys. This applies only to objects.

```ts
pipe(useCallValue())
  .pipe(exclude('foo'))
  .resolveSync({ foo: 123, bar: 'baz' }); // { bar: 'baz' }
```

###### hasProperty

Returns if an object has a owned property

```ts
pipe(useCallValue())
  .pipe(hasProperty('foo'))
  .resolveSync({ foo: new Date() }) // true
```

###### keys

Returns previous's value keys. Works only for objects

```ts
pipe(useCallValue())
  .pipe(keys())
  .resolveSync({ bar: 123, foo: new Date() }) // ['bar', 'foo']
```

###### maybe

Returns a property key value by a given path. This applies only to objects.

```ts
pipe(useCallValue())
  .pipe(maybe('foo.bar'))
  .resolveSync({ foo: { bar: 100 } }) // 100
```

You can provide a fallback value or a fallback pipeline if the path does not match the object schema.

If you start you pipeline with the `useCallValue` function, the monad will be invoked with an `undefined` value.

```ts
pipe(useCallValue())
  .pipe(maybe('foo.bar.baz', 123))
  .resolveSync({ foo: { bar: 100 } }) // 123

// or
const fallback = pipe(useCallValue());

pipe(useCallValue())
  .pipe(maybe('foo.bar.baz', fallback))
  .resolveSync({ foo: { bar: 100 } }) // undefined

// or
const fallback = pipe(put(10));

pipe(useCallValue())
  .pipe(maybe('foo.bar.baz', fallback))
  .resolveSync({ foo: { bar: 100 } }) // 10
```

###### merge

Merges the previous object with the given one

```ts
pipe(useCallValue())
  .pipe(merge({ foo: 'bar' }))
  .resolveSync({ bar: 'baz' }) // { foo: 'bar', bar: 'baz' }
```

###### pick

Returns a new object (previous value) with the given keys. This applies only to objects.

```ts
pipe(useCallValue())
  .pipe(pick('foo'))
  .resolveSync({ foo: 123, bar: 'baz' }); // { foo: 123 }
```

#### Misc functions

Misc functions are under the `misc` module.

```ts
import { misc } from 'effepi';
```

###### apply

Applies a pipeline using the async `resolve` method. 

Note: invoking a pipeline using this function with a sync method will throw an error.

```ts
const injectedPipeline = pipe(functions.useCallValue())
        .pipe(functions.add(10));

const testPipeline = pipe(functions.useCallValue())
  .pipe(functions.apply(injectedPipeline))
  .pipe(functions.multiplyBy(2)); 

testPipeline.resolve(2) // Promise(24)
```

###### applySync

Applies a pipeline using the sync `resolveSync` method.

Note: invoking a pipeline using this function with an async method will throw an error.

```ts
const injectedPipeline = pipe(functions.useCallValue())
        .pipe(functions.add(10));

const testPipeline = pipe(functions.useCallValue())
  .pipe(functions.applySync(injectedPipeline))
  .pipe(functions.multiplyBy(2)); 

testPipeline.resolve(2) // 24
```

###### put

Use this function to put a value at the beginning of the pipeline

```ts
pipe(put(10)).resolveSync(0) // 10
```

###### safeCall

Use this function to perform a safe function call (will not throw) with the previous value as argument

```ts
pipe(useCallValue())
  .pipe(safeCall(() => throw new Error('hello world')))
  .resolveSync(100) // will not throw, instead it will return 100
```

###### useCallValue

Use this function at the beginning of your pipeline to use the passed value to `resolve`/`resolveSync` and to function invokation

```ts
const p = pipe(useCallValue()).add(100);

p.resolve(200) // Promise(300)
p.resolveSync(10) // 110
p.toFunction()(123) // Promise(223)
p.toSyncFunction()(1000) // 1100
```

###### useValue

Use this function to return the previous pipeline value

```ts
pipe(useCallValue())
  .pipe(add(10))
  .pipe(useValue())
  .resolveSync(10) // 20
```

#### String functions

String functions are under the `string` module.

```ts
import { string } from 'effepi';
```

###### camelCase

Returns previous value in camel-case. Previous value must be a string.

```ts
pipe(useCallValue())
  .pipe(camelCase())
  .resolveSync('hello world') // 'helloWorld'
```

###### chars

Returns previous value as an array of chars. Previous value must be a string.

```ts
pipe(useCallValue())
  .pipe(chars())
  .resolveSync('hello') // returns ['h', 'e', 'l', 'l', 'o']
```

###### concat

Concatenate previous value with another string. Previous value must be a string.

```ts
pipe(useCallValue())
  .pipe(concat('world'))
  .resolveSync('hello') // 'helloworld'
```

###### includes

Returns if the previous value contains a portion of text. Previous value must be a string.

```ts
pipe(useCallValue())
  .pipe(includes('llo'))
  .resolveSync('hello') // true
```

###### length

Returns previous value length. Previous value must be a string.

```ts
pipe(useCallValue())
  .pipe(length())
  .resolveSync('hello') // 5
```

###### lowercase

Returns previous value in lower case. Previous value must be a string.

```ts
pipe(useCallValue())
  .pipe(lowercase())
  .resolveSync('HELLO') // 'hello'
```

###### pascalCase

Returns previous value in pascal-case. Previous value must be a string.

```ts
pipe(useCallValue())
  .pipe(pascalCase())
  .resolveSync('hello world') // 'HelloWorld'
```

###### repeat

Repeats previous value a number of times. Previous value must be a string.

```ts
pipe(useCallValue())
  .pipe(repeat())
  .resolveSync('hello') // hellohello

pipe(useCallValue())
  .pipe(repeat(2))
  .resolveSync('hello') // hellohellohello
```

###### replaceAll

Replaces all occurencies from the previous value. Previous value must be a string.

```ts
pipe(useCallValue())
  .pipe(replaceAll('l', '1'))
  .resolveSync('hello') // he110
```


CHANGELOG.md
## 0.1.1 (2019-06-01)


### Features

* adds apply context method ([8854e88](https://github.com/OctoD/effepi/commit/8854e88))
* adds apply/applyAsync functions ([32ec573](https://github.com/OctoD/effepi/commit/32ec573))
* adds camelCase ([1c2a3c9](https://github.com/OctoD/effepi/commit/1c2a3c9))
* adds changeSign, negative and positive fns ([f06ccf5](https://github.com/OctoD/effepi/commit/f06ccf5))
* adds chars function ([c8d7a9e](https://github.com/OctoD/effepi/commit/c8d7a9e))
* adds concat ([b223047](https://github.com/OctoD/effepi/commit/b223047))
* adds concat ([a802f29](https://github.com/OctoD/effepi/commit/a802f29))
* adds decrement and increment ([af04734](https://github.com/OctoD/effepi/commit/af04734))
* adds exactTypeOf ([881c227](https://github.com/OctoD/effepi/commit/881c227))
* adds exclude ([e48fd60](https://github.com/OctoD/effepi/commit/e48fd60))
* adds F & T ([7a5c41b](https://github.com/OctoD/effepi/commit/7a5c41b))
* adds filter ([265f528](https://github.com/OctoD/effepi/commit/265f528))
* adds filterWith ([f4080d3](https://github.com/OctoD/effepi/commit/f4080d3))
* adds find and findExact ([3f270ed](https://github.com/OctoD/effepi/commit/3f270ed))
* adds getTypeOf ([4506385](https://github.com/OctoD/effepi/commit/4506385))
* adds hasProperty ([9192e37](https://github.com/OctoD/effepi/commit/9192e37))
* adds ifElse ([c5ec626](https://github.com/OctoD/effepi/commit/c5ec626))
* adds includes ([bceef6a](https://github.com/OctoD/effepi/commit/bceef6a))
* adds join ([bc61b59](https://github.com/OctoD/effepi/commit/bc61b59))
* adds keys ([5d977cd](https://github.com/OctoD/effepi/commit/5d977cd))
* adds length ([fea1c20](https://github.com/OctoD/effepi/commit/fea1c20))
* adds length function ([c190d7e](https://github.com/OctoD/effepi/commit/c190d7e))
* adds lowercase and uppercase ([684553a](https://github.com/OctoD/effepi/commit/684553a))
* adds maybe monad ([d10aefd](https://github.com/OctoD/effepi/commit/d10aefd))
* adds memoize ([b239220](https://github.com/OctoD/effepi/commit/b239220))
* adds merge ([7005a43](https://github.com/OctoD/effepi/commit/7005a43))
* adds mutation delegate method to context ([8f3683e](https://github.com/OctoD/effepi/commit/8f3683e))
* adds nth ([7104c50](https://github.com/OctoD/effepi/commit/7104c50))
* adds ofType ([fc56cdf](https://github.com/OctoD/effepi/commit/fc56cdf))
* adds pascalCase ([4423c3a](https://github.com/OctoD/effepi/commit/4423c3a))
* adds pick function ([1038831](https://github.com/OctoD/effepi/commit/1038831))
* adds repeat ([9a4f80d](https://github.com/OctoD/effepi/commit/9a4f80d))
* adds replaceAll ([8e53a30](https://github.com/OctoD/effepi/commit/8e53a30))
* adds reverse ([dbcc45c](https://github.com/OctoD/effepi/commit/dbcc45c))
* adds takeBetween ([4d66f99](https://github.com/OctoD/effepi/commit/4d66f99))
* adds takeGreater and takeLower ([cb10a7f](https://github.com/OctoD/effepi/commit/cb10a7f))
* adds takeGreaterThan ([7b21078](https://github.com/OctoD/effepi/commit/7b21078))
* adds takeLowerThan ([0e93eec](https://github.com/OctoD/effepi/commit/0e93eec))
* adds takeOuter function ([f0f7612](https://github.com/OctoD/effepi/commit/f0f7612))
* adds toBinaryArray ([ece1989](https://github.com/OctoD/effepi/commit/ece1989))
* adds toBoolean and inverse ([4330cae](https://github.com/OctoD/effepi/commit/4330cae))
* adds toSet ([fedd66c](https://github.com/OctoD/effepi/commit/fedd66c))
* async/sync calls safeguards ([4bbf48d](https://github.com/OctoD/effepi/commit/4bbf48d))
* makes array functions stricter ([1d90b0b](https://github.com/OctoD/effepi/commit/1d90b0b))
* makes math functions stricter ([328268f](https://github.com/OctoD/effepi/commit/328268f))





src/math.ts
import { ExplicitCallable } from './pipe';
import { throwIfNotArray, throwIfNotNumber } from './helpers';

/**
 * Sums a value to the previous one.
 *
 * ```
 * pipe(useCallValue()).pipe(add(1)).resolveSync(10) // 11
 * ```
 *
 * @export
 * @param {number} value
 * @returns {ExplicitCallable<number, number>}
 */
export function add(value: number): ExplicitCallable<number, number> {
  return (arg: number) => {
    throwIfNotNumber('add', arg);
    return arg + value;
  };
}

/**
 * Changes previous value sign, from positive to negative and vice-versa.
 *
 * ```
 * pipe(put(-123)).pipe(changeSign()) // 123
 * ```
 *
 * @export
 * @returns {ExplicitCallable<number, number>}
 */
export function changeSign(): ExplicitCallable<number, number> {
  return (arg: number) => {
    throwIfNotNumber('changeSign', arg);
    return -arg;
  };
}

/**
 * Decrements the previous value by one.
 *
 * ```
 * pipe(useCallValue()).pipe(decrement()).resolve(44) // 41
 * ```
 *
 * @export
 * @returns {ExplicitCallable<number, number>}
 */
export function decrement(): ExplicitCallable<number, number> {
  return arg => {
    throwIfNotNumber('decrement', arg);
    return --arg;
  };
}

/**
 * Divides the previous value by the passed one.
 *
 * ```
 * pipe(useCallValue()).pipe(divideBy(2)).resolve(44) // 22
 * ```
 *
 * @export
 * @param {number} value
 * @returns {ExplicitCallable<number, number>}
 */
export function divideBy(value: number): ExplicitCallable<number, number> {
  return (arg: number) => {
    throwIfNotNumber('divideBy', arg);
    return arg / value;
  };
}

/**
 * Increments the previous value by one.
 *
 * ```
 * pipe(useCallValue()).pipe(increment()).resolve(44) // 45
 * ```
 *
 * @export
 * @returns {ExplicitCallable<number, number>}
 */
export function increment(): ExplicitCallable<number, number> {
  return arg => {
    throwIfNotNumber('increment', arg);
    return ++arg;
  };
}

/**
 * Multiplies the previous value by the given one
 *
 * ```
 * pipe(useCallValue()).pipe(multiplyBy(2)).resolve(44) // 88
 * ```
 *
 * @export
 * @param {number} value
 * @returns {ExplicitCallable<number, number>}
 */
export function multiplyBy(value: number): ExplicitCallable<number, number> {
  return (arg: number) => {
    throwIfNotNumber('multiplyBy', arg);
    return arg * value;
  };
}

/**
 * Converts previous' value from positive sign to negative unless is already negative.
 *
 * ```
 * pipe(useCallValue()).pipe(negative()).resolve(44) // -44
 * pipe(useCallValue()).pipe(negative()).resolve(-12) // -12
 * ```
 *
 * @export
 * @returns {ExplicitCallable<number, number>}
 */
export function negative(): ExplicitCallable<number, number> {
  return (arg: number) => {
    throwIfNotNumber('negative', arg);
    return arg > 0 ? -arg : arg;
  };
}

/**
 * Converts previous' value from negative sign to positive unless is already positive.
 *
 * ```
 * pipe(useCallValue()).pipe(positive()).resolve(44) // 44
 * pipe(useCallValue()).pipe(positive()).resolve(-12) // 12
 * ```
 *
 * @export
 * @returns {ExplicitCallable<number, number>}
 */
export function positive(): ExplicitCallable<number, number> {
  return (arg: number) => {
    throwIfNotNumber('positive', arg);
    return arg > 0 ? arg : -arg;
  };
}

/**
 * Raises to power the previous value by a given exponent
 *
 * ```
 * pipe(useCallValue()).pipe(pow(4)).resolve(2) // 16
 * pipe(useCallValue()).pipe(pow(2)).resolve(-2) // 4
 * ```
 *
 * @export
 * @param {number} exponent
 * @returns {ExplicitCallable<number, number>}
 */
export function pow(exponent: number): ExplicitCallable<number, number> {
  return (arg: number) => {
    throwIfNotNumber('pow', arg);
    return arg ** exponent;
  };
}

/**
 * Extracts the root of the previous value by a given exponent
 *
 * ```
 * pipe(useCallValue()).pipe(root(2)).resolve(4) // 2
 * pipe(useCallValue()).pipe(root(2)).resolve(9) // 3
 * ```
 *
 * @export
 * @param {number} exponent
 * @returns {ExplicitCallable<number, number>}
 */
export function root(exponent: number): ExplicitCallable<number, number> {
  return (arg: number) => {
    throwIfNotNumber('root', arg);
    return Math.pow(arg, 1 / exponent);
  };
}

/**
 * Subtracts the previous value by the given one
 *
 * ```
 * pipe(useCallValue()).pipe(subtract(2)).resolve(9) // 7
 * ```
 *
 * @export
 * @param {number} value
 * @returns {ExplicitCallable<number, number>}
 */
export function subtract(value: number): ExplicitCallable<number, number> {
  return (arg: number) => {
    throwIfNotNumber('subtract', arg);
    return arg - value;
  };
}

/**
 * Takes all numbers in a number array between two values (inclusive)
 *
 * ```
 * pipe(useCallValue()).pipe(takeBetween(5, 7)).resolve([4, 5, 6, 7, 8]) // [5, 6, 7]
 * ```
 *
 * @export
 * @param {number} start
 * @param {number} end
 * @returns {ExplicitCallable<number[], number[]>}
 */
export function takeBetween(start: number, end: number): ExplicitCallable<number[], number[]> {
  return (arg: number[]) => {
    throwIfNotArray(`takeBetween`, arg);

    const newArray: number[] = [];
    const length = arg.length;

    for (let i = 0; i < length; i++) {
      const value = arg[i];

      throwIfNotNumber('takeBetween', value);

      if (value >= start && value <= end) {
        newArray.push(value);
      }
    }

    return newArray;
  };
}

/**
 * Returns the greater number in a number array
 *
 * ```
 * pipe(useCallValue()).pipe(takeGreater()).resolve([4, 5, 44, 6, 7, 8]) // 44
 * ```
 *
 * @export
 * @returns {ExplicitCallable<number[], number>}
 */
export function takeGreater(): ExplicitCallable<number[], number> {
  return (arg: number[]) => {
    throwIfNotArray('takeGreater', arg);
    return Math.max.apply(Math, arg);
  };
}

/**
 * Takes all numbers in a number array greater than a given value.
 *
 * ```
 * pipe(useCallValue()).pipe(takeGreaterThan(8)).resolve([4, 5, 44, 6, 7, 8]) // [44]
 * ```
 *
 * This function accepts a second parameter (boolean) to include also the same value
 *
 * ```
 * pipe(useCallValue()).pipe(takeGreaterThan(8, true)).resolve([4, 5, 44, 6, 7, 8]) // [8, 44]
 * ```
 *
 * @export
 * @param {number} check
 * @param {boolean} [equal=false]
 * @returns {ExplicitCallable<number[], number[]>}
 */
export function takeGreaterThan(check: number, equal: boolean = false): ExplicitCallable<number[], number[]> {
  return (arg: number[]) => {
    throwIfNotArray(`takeGreaterThan`, arg);

    const newArray: number[] = [];
    const length = arg.length;

    for (let i = 0; i < length; i++) {
      const value = arg[i];

      throwIfNotNumber('takeGreaterThan', value);

      if (equal && value >= check) {
        newArray.push(value);
      } else if (value > check) {
        newArray.push(value);
      }
    }

    return newArray;
  };
}

/**
 * Returns the lower number in a number array
 *
 * ```
 * pipe(takeLower()).pipe(takeGreater()).resolve([4, 5, 44, 6, 7, 8]) // 4
 * ```
 *
 * @export
 * @returns {ExplicitCallable<number[], number>}
 */
export function takeLower(): ExplicitCallable<number[], number> {
  return (arg: number[]) => {
    throwIfNotArray('takeLower', arg);
    return Math.min.apply(Math, arg);
  };
}

/**
 * Takes all numbers in a number array lower than a given value.
 *
 * ```
 * pipe(useCallValue()).pipe(takeLowerThan(5)).resolve([4, 5, 44, 6, 7, 8]) // [4]
 * ```
 *
 * This function accepts a second parameter (boolean) to include also the same value
 *
 * ```
 * pipe(useCallValue()).pipe(takeLowerThan(8, true)).resolve([4, 5, 44, 6, 7, 8]) // [4, 5, 6, 7, 8]
 * ```
 *
 * @export
 * @param {number} check
 * @param {boolean} [equal=false]
 * @returns {ExplicitCallable<number[], number[]>}
 */
export function takeLowerThan(check: number, equal: boolean = false): ExplicitCallable<number[], number[]> {
  return (arg: number[]) => {
    throwIfNotArray(`takeLowerThan`, arg);

    const newArray: number[] = [];
    const length = arg.length;

    for (let i = 0; i < length; i++) {
      const value = arg[i];

      throwIfNotNumber('takeLowerThan', value);

      if (equal && value <= check) {
        newArray.push(value);
      } else if (value < check) {
        newArray.push(value);
      }
    }

    return newArray;
  };
}

/**
 * Takes all numbers in a number array lower than the first passed value or greater than the second passed value. Matching elements are discarded.
 *
 * ```
 * pipe(useCallValue()).pipe(takeOuter(5, 10)).resolveSync([3,4,5,10,11]) // [3, 4, 11]
 * ```
 *
 * @export
 * @param {number} start
 * @param {number} end
 * @returns {ExplicitCallable<number[], number[]>}
 */
export function takeOuter(start: number, end: number): ExplicitCallable<number[], number[]> {
  return (arg: number[]) => {
    throwIfNotArray(`takeOuter`, arg);

    const newArray: number[] = [];
    const length = arg.length;

    for (let i = 0; i < length; i++) {
      const value = arg[i];

      throwIfNotNumber('takeOuter', value);

      if (value < start || value > end) {
        newArray.push(value);
      }
    }

    return newArray;
  };
}
